### PR TITLE
[videoplayer/rbp] Add pi specific option to maintain vsync with pll adjustment

### PR DESCRIPTION
**Pull request by popcornmix**
New A/V sync option in system/audio to do "Adjust PLL".
This uses video clock (so perfect video syncing) but avoids having to resample
or drop/dupe audio packets which is normally required.

### CODE REVIEW

**Comment by popcornmix** [comment]
Ping @FernetMenta

File: addons/resource.language.en_gb/resources/strings.po [hunk 1]
```diff
@@ -18307,3 +18307,35 @@ msgstr ""
 msgctxt "#38041"
 msgid "Mixer"
 msgstr "" 
+
+#. Description of setting "System -> Audio Ouput -> A/V sync method" with label #38200
+#: system/settings/settings.xml
+msgctxt "#38200"
+msgid "PLL adustment to maintain audio/video sync"
```
Comment (anaconda) [inline addons/resource.language.en_gb/resources/strings.po]: Typo in `adustment`

File: xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.cpp [hunk 2]
```diff
@@ -2448,7 +2453,16 @@ CSampleBuffer* CActiveAE::SyncStream(CActiveAEStream *stream)
   if (!newerror || stream->m_syncState != CAESyncInfo::AESyncState::SYNC_INSYNC)
     return ret;
 
-  if (stream->m_resampleMode)
+  if (stream->m_pllAdjust > 0)  // pll adjust
+  {
+#if defined(TARGET_RASPBERRY_PI)
```
Comment (FernetMenta) [inline xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.cpp]: ifdefs should be avoided in platform agnostic code

File: xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.cpp [hunk 3]
```diff
@@ -2448,7 +2453,16 @@ CSampleBuffer* CActiveAE::SyncStream(CActiveAEStream *stream)
...
+    double e = std::max(std::min(error / 50.0, 1.0), -1.0);
+    double m_plladjust = 1.0 + e * stream->m_pllAdjust;
+    double m_last_plladjust = g_RBP.AdjustHDMIClock(m_plladjust);
```
Comment (FernetMenta) [inline xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.cpp]: there is always a better solution than a global

File: xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.cpp [hunk 4]
```diff
@@ -3300,13 +3314,14 @@ void CActiveAE::SetStreamResampleRatio(CActiveAEStream *stream, double ratio)
                                      &msg, sizeof(MsgStreamParameter));
 }
 
-void CActiveAE::SetStreamResampleMode(CActiveAEStream *stream, int mode)
+void CActiveAE::SetStreamResampleMode(CActiveAEStream *stream, int mode, float plladjust)
```
Comment (FernetMenta) [inline xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.cpp]: hmm, does not look like a clear design

File: xbmc/cores/VideoPlayer/VideoPlayerAudio.cpp [hunk 5]
```diff
@@ -102,6 +102,7 @@ bool CVideoPlayerAudio::OpenStream(CDVDStreamInfo &hints)
   bool allowpassthrough = !CSettings::GetInstance().GetBool(CSettings::SETTING_VIDEOPLAYER_USEDISPLAYASCLOCK);
   if (hints.realtime)
     allowpassthrough = false;
+  allowpassthrough |= CSettings::GetInstance().GetInt("audiooutput.plladjust") > 0;
```
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/VideoPlayerAudio.cpp]: nothing outside of AE is allowed to access AE settings

**Comment by FernetMenta** [comment]
Why don't you instantiate a special kind of resampler for this this mode?

File: addons/resource.language.en_gb/resources/strings.po [hunk 6]
```diff
@@ -18307,3 +18307,35 @@ msgstr ""
...
+msgstr ""
+
+#. Description of setting "Videos -> Playback -> A/V sync method" with label #38201
+#: system/settings/settings.xml
+msgctxt "#38201"
+msgid "Allows sync adjustment without resampling. Lower the settings if you get audio/video dropouts."
+msgstr ""
+
+msgctxt "#38202"
+msgid "Off"
```
Comment (da-anda) [inline addons/resource.language.en_gb/resources/strings.po]: we already have 6 labels for "off" - mind checking if one of these shares the same context and can be reused?

File: addons/resource.language.en_gb/resources/strings.po [hunk 7]
```diff
@@ -18307,3 +18307,35 @@ msgstr ""
 msgctxt "#38041"
 msgid "Mixer"
 msgstr "" 
+
+#. Description of setting "System -> Audio Ouput -> A/V sync method" with label #38200
```
Comment (da-anda) [inline addons/resource.language.en_gb/resources/strings.po]: that contextual hint is actually wrong, as this is the label of a setting and not it's description :) Also, will users know what "PLL adjustment" means? Could this be phrased less technical as I really doubt the average user knows what PLL is (even I don't know - I only know it's somehow related to playback)

File: addons/resource.language.en_gb/resources/strings.po [hunk 8]
```diff
@@ -18307,3 +18307,35 @@ msgstr ""
...
+#: system/settings/settings.xml
+msgctxt "#38200"
+msgid "PLL adustment to maintain audio/video sync"
+msgstr ""
+
+#. Description of setting "Videos -> Playback -> A/V sync method" with label #38201
+#: system/settings/settings.xml
+msgctxt "#38201"
+msgid "Allows sync adjustment without resampling. Lower the settings if you get audio/video dropouts."
```
Comment (da-anda) [inline addons/resource.language.en_gb/resources/strings.po]: as there are different values possible, shouldn't we describe which effect a lower and a higher value will have? Like, the higher the more delay (or whatever happens)

**Comment by MartijnKaijser** [comment]
I assume this is quite outdated and maybe obsolete by now.

**Pull request closed without merging**